app/file.go

```diff
@@ -623,32 +623,46 @@ func (a *App) UploadFileX(channelId, name string, input io.Reader,
 		}
 	}
 
-	aerr = t.readAll()
+	written, aerr := t.writeFile(io.MultiReader(t.buf, t.limitedInput), t.fileinfo.Path)
 	if aerr != nil {
-		return t.fileinfo, aerr
+		return nil, aerr
 	}
 
-	aerr = t.runPlugins()
+	if written > t.maxFileSize {
+		a.RemoveFile(t.fileinfo.Path)
+		return nil, t.newAppError("api.file.upload_file.too_large_detailed.app_error",
+			"", http.StatusRequestEntityTooLarge, "Length", t.ContentLength, "Limit", t.maxFileSize)
+	}
+
+	t.fileinfo.Size = written
+
+	file, aerr := a.FileReader(t.fileinfo.Path)
 	if aerr != nil {
-		return t.fileinfo, aerr
+		return nil, aerr
+	}
+	defer file.Close()
+
+	aerr = a.runPluginsHook(t.fileinfo, file)
+	if aerr != nil {
+		return nil, aerr
 	}
 
 	// Concurrently upload and update DB, and post-process the image.
 	wg := sync.WaitGroup{}
 
-	if !t.Raw && t.fileinfo.IsImage() {
+	if !t.Raw && t.fileinfo.IsImage() && t.fileinfo.HasPreviewImage {
+		file, aerr = a.FileReader(t.fileinfo.Path)
```

Comment:
Can you clarify why we are getting a fileReader again here after L639?

Reply:
Because `a.runPluginsHook()` can potentially read and change the file.